specs/core/0_beacon-chain.md

```diff
@@ -1886,135 +1949,301 @@ def process_transfer(state: BeaconState, transfer: Transfer) -> None:
+            deltas[index] += get_base_reward(state, index) * boundary_attesting_balance // total_balance
+        else:
+            deltas[index] -= get_base_reward(state, index)
+        # Expected head
+        if index in get_attesting_indices(matching_head_attestations):
+            deltas[index] += get_base_reward(state, index) * matching_head_balance // total_balance
+        else:
+            deltas[index] -= get_base_reward(state, index)    
+        # Proposer bonus
+        proposer_index = get_beacon_proposer_index(state, inclusion_slot(state, index))
+        deltas[proposer_index] += base_reward(state, index) // ATTESTATION_INCLUSION_REWARD_QUOTIENT
+```
+
+When blocks are not finalizing normally...
+            
+```python
+def compute_inactivity_leak_deltas(state: BeaconState) -> List[Gwei]:
+    deltas = [0 for index in range(len(state.validator_registry))]
+    boundary_attestations = get_previous_epoch_boundary_attestations(state)
+    matching_head_attestations = get_previous_epoch_matching_head_attestations(state)
+    active_validator_indices = get_active_validator_indices(state.validator_registry, get_previous_epoch(state))
+    epochs_since_finality = get_current_epoch(state) + 1 - state.finalized_epoch
+    for index in active_validator_indices:
+        if index not in get_attesting_indices(get_previous_epoch_attestations(state)):
+            deltas[index] -= get_inactivity_penalty(state, index, epochs_since_finality)
+        else:
+            # If a validator did attest, apply a small penalty for getting attestations included late
+            deltas[index] += (
+                base_reward(state, index) * MIN_ATTESTATION_INCLUSION_DELAY //
+                inclusion_distance(state, index)
+            )
+            deltas[index] -= base_reward(state, index)
+                    
+        if index not in get_attesting_indices(boundary_attestations):
+            deltas[index] -= get_inactivity_penalty(state, index, epochs_since_finality)
+        if index not in get_attesting_indices(matching_head_attestations):
+            deltas[index] -= get_base_reward(state, index)
+    # Penalize slashed-but-inactive validators as though they were active but offline
+    for index in range(len(state.validator_registry)):
+        if index not in active_validator_indices and state.validator_registry[index].slashed:
+            deltas[index] -= (
+                2 * get_inactivity_penalty(state, index, epochs_since_finality) +
+                get_base_reward(state, index)
+            )
+    return deltas
+```
 
-##### Attestation inclusion
+##### Crosslinks
 
-For each `index` in `previous_epoch_attester_indices`, we determine the proposer `proposer_index = get_beacon_proposer_index(state, inclusion_slot(state, index))` and set `state.validator_balances[proposer_index] += base_reward(state, index) // ATTESTATION_INCLUSION_REWARD_QUOTIENT`.
+```python
+def get_crosslink_deltas(state: BeaconState) -> List[Gwei]:
```

Comment:
Agreed with @hwwhww to return a Tuple of lists of gwei (the positive and negative deltas in separate lists)

Reply:
Done